### PR TITLE
New set of DevOps scripts

### PR DESCRIPTION
**Pull request by Greg-Szabo**
Ansible was refactored to use the packaged applications. This way ansible is used for what's it's best for: configuration management. It's also much less complex than before. Easier to understand or reuse.
Package creation scripts were added (Debian and CentOS).
Terraform was extended to create DNS names in AWS for our public testnets.
Our Jenkins/ChatOps infrastructure directly use these scripts, they are well tested.

### CODE REVIEW

File: ansible/app_options_files/empty [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (melekes) [inline ansible/app_options_files/empty]: looks strange. why we need it?

File: ansible/app_options_files/public_testnet [hunk 2]
```diff
@@ -0,0 +1,132 @@
+    "accounts": [
```
Comment (melekes) [inline ansible/app_options_files/public_testnet]: should this be public?

File: ansible/roles/config/templates/genesis-server.json.j2 [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  "validators":
+  [
+{% if validators == true %}
+{% set comma = joiner(",") %}
+{% for host in groups[testnet_name] %}
+    {{ comma() }}
+    {
+      "pub_key": {
+        "data": "{{hostvars[host]["pubkey"]["data"]}}",
+        "type": "ed25519"
```
Comment (melekes) [inline ansible/roles/config/templates/genesis-server.json.j2]: `"{{hostvars[host]["pubkey"]["type"]}}"`

File: devops/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# DevOps tools
```
Comment (melekes) [inline devops/README.md]: do we wanna open source that?

File: devops/rpmbuild/extrafiles/basecoin/usr/share/basecoin/key2.json [hunk 5]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (melekes) [inline devops/rpmbuild/extrafiles/basecoin/usr/share/basecoin/key2.json]: should this be a part of rpm?

File: devops/rpmbuild/Makefile [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+	cat spectemplates/$*.spec >> SPECS/$*.spec
+
+package-rpm-%: prepare-spec-%
+	@echo "*** Packaging RPM $* version $($*_version)"
+	rpmbuild -bb SPECS/$*.spec
+	./sign RPMS/x86_64/$*-$($*_version)-$(BUILD_NUMBER).x86_64.rpm
+	rpm -Kv RPMS/x86_64/$*-$($*_version)-$(BUILD_NUMBER).x86_64.rpm
+	@echo "*** Packaged RPM $* version $($*_version)"
+
+package-deb-%: version-%
```
Comment (melekes) [inline devops/rpmbuild/Makefile]: wish there were a tool which does the same job for us. so much code to support

File: docker/README.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
+# Docker container description for Tendermint applications
```
Comment (melekes) [inline docker/README.md]: huh? why do we need all docker files in one place?

File: ansible/app_options_files/empty (hunk 1 above)
Comment (Greg-Szabo) [inline ansible/app_options_files/empty]: Fair question. :) When you want to deploy a new basecoin infrastructure, you can add different sets of wallets up front. The config file stores them under app_options as a dict. Ansible has an input parameter that chooses the file from the app_options_files directory for the wallets during configuration. If there's no file, ansible errors out. Hence if you want ansible to create an empty testnet (with no wallets) you still have to have a file in this folder.
I'm considering refacting this completely, because later it turned out that the ethermint app_options part is not compatible with basecoin. This means that the content of this folder can only be used for basecoin and not much else...

File: ansible/app_options_files/public_testnet (hunk 2 above)
Comment (Greg-Szabo) [inline ansible/app_options_files/public_testnet]: So far I didn't see any problems with it. This contains public wallet addresses. No private keys are stored. People can check what kind of coins we have in our network, but as soon as bucky and the others start spending it, it's going to be a different amount.

File: ansible/app_options_files/empty (hunk 1 above)
Comment (melekes) [inline ansible/app_options_files/empty]: > I'm considering refacting this completely, because later it turned out that the ethermint app_options part is not compatible with basecoin.

sorry about that. don't think we should put a lot of stuff in here. don't you think it is better to create a separate playbooks/folders for ethermint/basecoin? like `ansible-basecoin`, `ansible-ethermint`?

File: ansible/roles/config/templates/genesis-server.json.j2 (hunk 3 above)
Comment (Greg-Szabo) [inline ansible/roles/config/templates/genesis-server.json.j2]: nice catch, thanks. Fixed.

File: ansible/app_options_files/empty (hunk 1 above)
Comment (melekes) [inline ansible/app_options_files/empty]: just an idea. not for this PR of course

File: devops/README.md (hunk 4 above)
Comment (Greg-Szabo) [inline devops/README.md]: I think we can. It might help people who want to set up testnets. Anything that we wanted private (keys, some of the infrastructure elements) should move to a private repo - as soon as we can open another private repo.

File: devops/rpmbuild/extrafiles/basecoin/usr/share/basecoin/key2.json (hunk 5 above)
Comment (Greg-Szabo) [inline devops/rpmbuild/extrafiles/basecoin/usr/share/basecoin/key2.json]: This is something that we should discuss. The key2.json is generated if you run basecoin init but it's a static file. I don't think the binary should generate static files for us, I think it should be supplied separately - in this case the RPM (or deb). The other issue is that the key file is generated into the configuration directory but a key file is more of a personal asset then a configuration for the server. Right now this is in limbo:
1. either I remove the key file from the repo and keep things as is, or
2. I remove the file but we put it into the basecoin repo so the rpm can pull it
3. keep the file.

I don't like the third option, all necessary files for packaging should be part of the source repository. I don't like the first option either because we clutter the configuration folder with unnecessary files.

But the RPM should definitely have this file, preferably installing it out of the way, like in /usr/share/basecoin (as any other application in linux that has extra examples).

File: ansible/README.md [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+sudo pip install ansible
+```
+
+To make life easier, you can start an SSH Agent and load your SSH key(s). This way ansible will have an uninterrupted way of connecting to your servers.
+
+```
+ssh-agent > ~/.ssh/ssh.env
+source ~/.ssh/ssh.env
+
+ssh-add private.key
```
Comment (adrianbrink) [inline ansible/README.md]: What is contained in the `ssh.env` file?

Also for macOS the command is: `ssh-add -K ~/.ssh/id_rsa` to add it the ssh-agent.

File: docker/README.md (hunk 7 above)
Comment (Greg-Szabo) [inline docker/README.md]: I want to separate building binaries and packaging them for multiple reasons.
1. The build process is a development process requiring a special environment set up. If someone wants to install the baked applications they shouldn't need to install the go language for that. This is what packages are for.
2. I consider a docker container a package type and this way all package building is coded in this repository.
3. This way, the automated package builder (bithead) can build docker packages separated from the build process. Since the build system is undergoing rapid changes (it's still being built), it's easier to change the docker packaging method in the tools repository than in tendermint/basecoin/ethermint repositories.
4. This way the tendermint/basecoin/ethermint repositories can focus on the code and the building method of the binary. If a new package design emerges (say docker++), it doesn't affect these repositories.
5. Most importantly, with the introduction of OpenShift I needed an automatic way of building and uploading docker containers using bithead. Since all code for bithead is in this repo, including the mintnet-kubernetes tool, it made sense to move all docker-related tools in one place.

File: ansible/README.md [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+* reset.yml - Stop the application, reset the configuration and data, then start the application again. You need to pass `-e service=<servicename>`, like `-e service=basecoin`. It will restart the underlying tendermint application too.
+* restart.yml - Restart a service on all nodes. You need to pass `-e service=<servicename>`, like `-e service=basecoin`. It will restart the underlying tendermint application too.
+* stop.yml - Stop the application. You need to pass `-e service=<servicename>`.
+* start.yml - Start the application. You need to pass `-e service=<servicename>`.
+
+The roles are self-sufficient under the `roles/` folder.
+
+* install-tendermint - install the tendermint application. It can install release packages or custom-compiled binaries.
+* install - install the application defined in the `service` parameter. It can install release packages and update them with custom-compiled binaries.
+* rmdb - delete the tendermint database for a service.
```
Comment (adrianbrink) [inline ansible/README.md]: Does this also remove the ABCI app database? If it doesn't it would be very useful if it would. So something that resets only the data but leaves the configuration alone.

File: ansible/app_options_files/empty (hunk 1 above)
Comment (adrianbrink) [inline ansible/app_options_files/empty]: @melekes @Greg-Szabo Having different playbooks is probably the way forward. Ethermint and Basecoin will not converge to be exactly the same. Furthermore once we add another zone that configuration will most likely also be different. I am thinking like the bitcoin zone or any other zone.

File: devops/rpmbuild/extrafiles/basecoin/usr/share/basecoin/key2.json (hunk 5 above)
Comment (adrianbrink) [inline devops/rpmbuild/extrafiles/basecoin/usr/share/basecoin/key2.json]: What about renaming the key2.json to test-key-2.json, just to make it very clear that they are test files.

File: docker/README.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+* Run the tendermint container and expose the ports that allow clients to connect. The --proxy_app should contain the basecoin application's IP address and port.
+```
+docker run --rm -d -v basecoindata:/basecoin basecoin start --without-tendermint
+docker run --rm -d -v data:/tendermint -p 46656-46657:46656-46657 tendermint node --proxy_app tcp://172.17.0.2:46658
+```
+
+## Ethermint
+
+Build the container:
+Copy the `ethermint` binary and the setup folder to the `ethermint` folder.
```
Comment (adrianbrink) [inline docker/README.md]: Where does the binary come from? Is this using the docker images that we publish to docker hub with every release?

File: docker/README.md [hunk 11]
```diff
@@ -0,0 +1,95 @@
+# Docker container description for Tendermint applications
+
+* Overview (#Overview)
+* Tendermint (#Tendermint)
+* Basecoin (#Basecoin)
+* Ethermint (#Ethermint)
+
+## Overview
```
Comment (adrianbrink) [inline docker/README.md]: Maybe it would be cool to extend this with `docker-compose.yaml` files to orchestrate this properly.

File: ansible/README.md [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+Optional for DigitalOcean droplets:
+* DigitalOcean API Token
+* python dopy package
+
+Head over to the [Terraform folder](https://github.com/tendermint/tools/tree/master/terraform-digitalocean) for a description on how to get a DigitalOcean API Token.
+
+Optional for Amazon AWS instances:
+* Amazon AWS API access key ID and secret access key.
+
+The cloud inventory scripts come from the ansible team at their [GitHub](https://github.com/ansible/ansible) page. You can get the latest version from the contrib/inventory folder.
```
Comment (Greg-Szabo) [inline ansible/README.md]: Cloud inventory scripts are dynamic scripts (usually in Python) that can query a specific cloud infrastructure for the list of nodes running. They are in the inventory folder but only the Digital Ocean script was copied over for now. If you run the script, it's going to give you a list of nodes running on Digital Ocean with your account (you have to provide the API key). Ansible can directly read this output and apply the changes on the nodes as it was an ansible inventory.

File: ansible/README.md [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+CentOS/RedHat:
+```
+sudo yum install epel-release
+sudo yum install ansible
+```
+
+Mac OSX:
+```
+sudo easy_install pip
+sudo pip install ansible
```
Comment (Greg-Szabo) [inline ansible/README.md]: I think I had an issue with that, although I can't remember what it was. The Ansible documentation states that: "The preferred way to install ansible on a Mac is via pip." [here](http://docs.ansible.com/ansible/intro_installation.html#latest-releases-on-mac-osx).
Did `brew install ansible` work for you? What version did it install? If it's at least 2.3.0 then we can add it.

File: ansible/README.md (hunk 8 above)
Comment (Greg-Szabo) [inline ansible/README.md]: Is `-K` mandatory on the mac? If yes, then it's a problem as the linux version doesn't have that option.
When you run ssh-agent, it blurts out a few lines that allows you to connect to the same instance of ssh-agent from a different console. If you save those lines (like I did in line 56), you can log on to the machine and invoke that file and you will have the keys ready to use. So, effectively you only need to add your private keys to the ssh-agent once every reboot.

File: ansible/README.md (hunk 9 above)
Comment (Greg-Szabo) [inline ansible/README.md]: Ah, thanks for catching that, the rmdb.yml is not necessary any more, it was removed. Use reset.yml if you need to reset the database which will reset both tendermint and the service (or ABCI app). As it turns out, unsafe_reset_all (which is what reset.yml does) does not remove the private/public key and it does not change the genesis.json or config.toml files. It removes the databases and it resets the counters in priv_validator.json.
BTW, you pointed me to another bug, that I'm going to fix now...

File: docker/README.md (hunk 11 above)
Comment (Greg-Szabo) [inline docker/README.md]: Yes, that's one of the next steps I'm working on, I just didn't get there yet. :)

File: ansible/README.md (hunk 13 above)
Comment (adrianbrink) [inline ansible/README.md]:
```➜  .ssh ansible --version
ansible 2.3.1.0
  config file =
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Apr  4 2017, 08:47:57) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)]```

File: ansible/app_options_files/public_testnet (hunk 2 above)
Comment (melekes) [inline ansible/app_options_files/public_testnet]: ok, cool

File: ansible/README.md (hunk 8 above)
Comment (adrianbrink) [inline ansible/README.md]: On macOS it is mandatory because otherwise it won't get added persistently and you will have to run it again after every reboot.

File: devops/rpmbuild/extrafiles/basecoin/usr/share/basecoin/key2.json (hunk 5 above)
Comment (melekes) [inline devops/rpmbuild/extrafiles/basecoin/usr/share/basecoin/key2.json]: correct me if I am wrong, but the key file must be generated on a node. or this is like an example for tutorials? I am not getting it

File: docker/README.md (hunk 10 above)
Comment (Greg-Szabo) [inline docker/README.md]: This is a separate effort to automate the docker container creation and upload. The binary comes from a separate GO build that happens earlier (Jenkins does that) and it can be provided to the Dockerfile as an input argument. For example:
`docker build -t ethermint --build-arg ETHERMINT_BINARY=$GOPATH/bin/ethermint ethermint`
Right now it seems that the `--build-arg` option will only accept paths from the sub-folder where the Dockerfile is. Hence if you copy it there, you can use `--build-arg ETHERMINT_BINARY=./ethermint` which is the default setting so you can omit it. Same with `--build-arg SETUP_DIR=setup`.
This is going to be part of the packaging process, just like the RPM or DEB creation.

File: docker/README.md (hunk 7 above)
Comment (melekes) [inline docker/README.md]: ok, makes sense. thanks!

File: ansible/README.md (hunk 13 above)
Comment (Greg-Szabo) [inline ansible/README.md]: Ok, according to blogs on the Internet ansible should be packaged with brew. (I don't have a Mac with brew lying around right now.) I'm going to assume that it's a recent version. I added it to the readme.

File: ansible/README.md (hunk 8 above)
Comment (Greg-Szabo) [inline ansible/README.md]: Ok, so according to the (developer's documentation)[https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man1/ssh-add.1.html] the `-K` option will add your passphrase to the keychain. This is a Mac-specific setting and it's not necessary for ssh-agent. (It seems that Apple is trying to work around ssh-agent completely. It seems a weird idea.) I think this should not be added to the documentation as it's confusing and not compatible with all systems.
You should be able to add the key to ssh-agent without the `-K` option on a Mac.

File: devops/rpmbuild/extrafiles/basecoin/usr/share/basecoin/key2.json (hunk 5 above)
Comment (Greg-Szabo) [inline devops/rpmbuild/extrafiles/basecoin/usr/share/basecoin/key2.json]: This is an example for tutorials. I didn't want to change the name (although I agree, test-key-2.json would be less confusing) because this is how it is mentioned in the tutorials.
The key file (priv_validator.json) is generated dynamically. :)

File: ansible/app_options_files/empty (hunk 1 above)
Comment (Greg-Szabo) [inline ansible/app_options_files/empty]: Agreed, although I was thinking separate genesis.json files for different applications. This is actually quite well managed in ethermint: the startup genesis.json is part of the source code repository and the package just puts it in place. Maybe we should do something similar with basecoin: have multiple genesis.json files pre-generated as examples and during installation the right one is copied over to the config folder. I don't really see the reason to manage genesis.json files for all kinds of applications here.

File: ansible/README.md (hunk 8 above)
Comment (Greg-Szabo) [inline ansible/README.md]: Hm..., let me rephrase that once more, I think Apple is missing a point here. The ssh-agent's job is to cache your private key in memory for security and convenience purposes. The workflow is the following:
1. During bootup you provide a passphrase manually so the private key can be read and decrypted from the file system into memory.
2. You make sure your environment is pointing to the ssh-agent using environment variables so if any processes (ssh for example) needs authenticating, they can request the ssh-agent to sign stuff that the process presents. This way the private key is secured with encryption on the file system but still available for signing through ssh-agent. Note that the ssh-agent will not provide the private key to the process, but instead it will receive bytes from the process, sign them with the private key and send it back to the process. So the private key is also safe in memory. (As long as the kernel is doing it's job properly.)
This is more convenient than manually entering a passphrase every time SSH (or a similar process) requires authentication, as well as more secure because the process doesn't get direct access to the private key.

Then comes Apple and cuts through this. The point of the workflow above was that in security there is always a manual point somewhere that initiates the whole security algorithm (consensus for symmetric secret) but ssh-agent allows you to go through this only once (per reboot cycle) and have the convenience of better security and less interaction for the system. Now Apple adds a feature (`-K`) which is not compatible with any other Unix or Linux system but allows to connect to yet another Apple-only system (keychain), to store the passphrase centrally and sync it to a cloud service - managed by Apple. So, until now you had to hack the in-memory ssh-agent to read the private key from memory, now you have multiple attack vectors: you can hack the keychain, the Apple cloud service or the ssh-agent memory. Also, since you still have to run ssh-agent at every reboot, you can try to hack into the ssh-agent - keychain communication. (I didn't check, but it would be interesting to see how secure that is.) By adding another layer of abstraction they made something quite secure into a nightmare. All parts of your ssh private key is stored by a digital system now.
Note that if you think about it, it is more secure to store your private key as plain text without encryption on your encrypted Apple hard-drive because that will not end up in the cloud. (Well, except if you turned on the feature to upload your folders to iCloud drive.) I know this is exaggeration, but it has its points.
Admittedly linux has features like HashiCorp Vault, to store multiple secrets in one place, but there at least you manage all the data and not a third-party with dubious license agreements.
So, everything summarized, ssh-agent/ssh-add requires you to enter the passphrase at each boot for a reason and I'm against using `-K` in OSX.

With that said, I've added a note to the readme so people know about this option for OSX.

File: docker/README.md (hunk 11 above)
Comment (antoineherzog) [inline docker/README.md]: We did this for the hackethon. You can have a look. We will finalize it tomorow:
https://github.com/multiverseHQ/deployment/tree/master/docker_compose

File: docker/README.md (hunk 10 above)
Comment (adrianbrink) [inline docker/README.md]: Mh, are we centralising the building of docker images into this repo or is everyone still pushing their own docker images. 

I think there should only one place where we build the images and everywhere else we pull them.

But this isn't a blocker for me.

**Comment by zramsay** [comment]
what's this blocking on?
